docs/design/specs/runtime-async.md

```diff
@@ -46,13 +46,11 @@ Async methods support the following suspension points:
 
 Each of the above methods will have semantics analogous to the current AsyncTaskMethodBuilder.AwaitOnCompleted/AwaitUnsafeOnCompleted methods. After calling this method, it can be presumed that the task has completed.
 
-Only local variables which are "hoisted" may be used across suspension points. That is, only "hoisted" local variables will have their state preserved after returning from a suspension. On methods with the `localsinit` flag set, non-"hoisted" local variables will be initialized to their default value when resuming from suspension. Otherwise, these variables will have an undefined value. To identify "hoisted" local variables, they must have an optional custom modifier to the `System.Runtime.CompilerServices.HoistedLocal` class, which will be a new .NET runtime API. This custom modifier must be the last custom modifier on the variable. It is invalid for by-ref variables, or variables with a by-ref-like type, to be marked hoisted. Hoisted local variables are stored in managed memory and cannot be converted to unmanaged pointers without explicit pinning.
-The code generator is free to ignore the `HoistedLocal` modifier if it can prove that this makes no observable difference in the execution of the generated program. This can be observable in diagnostics since it may mean the value of a local with the `HoistedLocal` modifier will not be available after certain suspension points.
+Local variables used across suspension points are considered "hoisted." That is, only "hoisted" local variables will have their state preserved after returning from a suspension. By-ref variables may not be hoisted across suspension points, and any read of a by-ref variable after a suspension point will produce null.
 
 Async methods have some temporary restrictions with may be lifted later:
 * The `tail` prefix is forbidden
 * Usage of the `localloc` instruction is forbidden
-* Pinning locals may not be marked `HoistedLocal`
```

Comment:
Should we have a similar rule that locals marked pinned are zeroed at suspension points?

Reply:
Yes. Lets break pinned locals in the same way as byrefs.

Reply:
Typically a pinned local is an internal pointer (`&`), so it is already broken by the other rule. 

But object references (i.e. `o`) can be pinned as well (some forms of `fixed(..){..}` actually use that), so we need a separate rule for this case.